### PR TITLE
fix: stmgr: make the tipset and height agree when estimating gas

### PR DESCRIPTION
**Pull request by Stebalien**
## Related Issues

Should fix https://github.com/filecoin-project/ref-fvm/issues/1648.

## Proposed Changes

Specifically re-execute all messages in the current tipset, tacking the new message onto the end. That way, the epoch is the epoch of the current tipset.

We could try to "make" a fake block and use that, but that's unlikely to work well.

## Checklist

Before you mark the PR ready for review, please make sure that:

- [x] Commits have a clear commit message.
- [x] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [x] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

**Comment by maciejwitowski** [comment]
Tested https://github.com/filecoin-project/lotus/pull/10216, it works.

**Comment by snissn** [comment]
This test also confirms this PR works https://github.com/filecoin-project/lotus/pull/10206

File: chain/stmgr/call.go [hunk 1]
```diff
@@ -142,14 +136,14 @@ func (sm *StateManager) callInternal(ctx context.Context, msg *types.Message, pr
 	buffStore := blockstore.NewTieredBstore(sm.cs.StateBlockstore(), blockstore.NewMemorySync())
 	vmopt := &vm.VMOpts{
 		StateBase:      stateCid,
-		Epoch:          vmHeight,
+		Epoch:          ts.Height(),
```
Comment (snissn) [inline chain/stmgr/call.go:139]: we're changing the value of Epoch here from vmHeight (ts.Height()+1) to ts.Height() is that off by 1 change ok?

File: chain/stmgr/call.go [hunk 2]
```diff
@@ -107,22 +107,16 @@ func (sm *StateManager) callInternal(ctx context.Context, msg *types.Message, pr
...
+	// Unless executing on a specific state cid, apply all the messages from the current tipset
+	// first. Unfortunately, we can't just execute the tipset, because that will run cron. We
+	// don't want to apply miner messages after cron runs in a given epoch.
+	if stateCid == cid.Undef {
+		stateCid = ts.ParentState()
+		tsMsgs, err := sm.cs.MessagesForTipset(ctx, ts)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to lookup messages for parent tipset: %w", err)
 		}
+		priorMsgs = append(tsMsgs, priorMsgs...)
```
Comment (magik6k) [inline chain/stmgr/call.go:119]: Are we sure this isn't breaking StateCall / `lotus state compute-state`?

Exchanges rely on this to track transactions.

File: chain/stmgr/call.go (hunk 1 above)
Comment (Stebalien) [reply chain/stmgr/call.go:139]: Yes, that's the point.

File: chain/stmgr/call.go (hunk 2 above)
Comment (Stebalien) [reply chain/stmgr/call.go:119]: That uses https://github.com/filecoin-project/lotus/blob/7aa6663127982aa404e68bdefaadb3f041eb5120/chain/stmgr/utils.go#L68

Comment (Stebalien) [reply chain/stmgr/call.go:119]: Which is also incorrect? Well, it's correct if you don't pass any messages.

Comment (Stebalien) [reply chain/stmgr/call.go:119]: Hm. It does affect `StateCall`. I'm going to rework this a bit.

Comment (Stebalien) [reply chain/stmgr/call.go]: Done. I now only do this for `CallWithGas`.

**Review by arajasek — APPROVED** [review]

**Pull request merged**